src/aws/nodemanager/nodemanager.py

```diff
@@ -113,33 +139,52 @@ def generate_heartbeat(self, notify=True):
     def process_heartbeat(self, hb, source) -> Packet:
         hb['source'] = source  # Set the source IP of the heartbeat (i.e., the worker).
         self.notify(hb)  # Forward the heartbeat to the monitor for metrics.
-        if len(self._tasks)>0:  #If there are tasks in the taskpool send a new command to the worker
-            packet=CommandPacket(command="task")
-            current_task=self._tasks.pop(0)
-            packet["task_data"]= current_task.get_task_data()
-            current_task.state=TaskState.RUNNING
-            self._tasks_running.append(current_task)
+
+        if hb['instance_id'] in self._task_assignment and \
+                not hb['instance_id'] in self._tasks_running:
+            packet = CommandPacket(command="task")
+            packet['task'] = self._task_assignment[hb['instance_id']].pop(0)
+            self._task_processing[hb['instance_id']] = [packet['task']]
+            return packet
+
+        # When a new instance is available
+        elif not hb['instance_id'] in self._task_assignment.keys():
```

Comment:
A call to `keys()` is not needed here. You can simply do `in self_task_assignment`